<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Исходные-данные" data-toc-modified-id="Исходные-данные-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Исходные данные</a></span></li><li><span><a href="#Анализ-таблиц" data-toc-modified-id="Анализ-таблиц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Анализ таблиц</a></span></li><li><span><a href="#Задания" data-toc-modified-id="Задания-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задания</a></span><ul class="toc-item"><li><span><a href="#Задача-№1" data-toc-modified-id="Задача-№1-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Задача №1</a></span></li><li><span><a href="#Задача-№2" data-toc-modified-id="Задача-№2-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Задача №2</a></span></li><li><span><a href="#Задача-№3" data-toc-modified-id="Задача-№3-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Задача №3</a></span></li><li><span><a href="#Задача-№4" data-toc-modified-id="Задача-№4-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Задача №4</a></span></li><li><span><a href="#Задача-№5" data-toc-modified-id="Задача-№5-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Задача №5</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Задача-№6" data-toc-modified-id="Задача-№6-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Задача №6</a></span></li><li><span><a href="#Задача-№7" data-toc-modified-id="Задача-№7-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Задача №7</a></span></li></ul></div>

# База данных книжного сервиса

__Заказчик__

Коммерческий отдел

__Цель исследования__

Анализ базы данных нового продукта для создания новых предложений

## Исходные данные

Таблица `books`

Содержит данные о книгах:


* `book_id` — идентификатор книги;
* `author_id` — идентификатор автора;
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.

Таблица `authors`

Содержит данные об авторах:


* `author_id` — идентификатор автора;
* `author` — имя автора.

Таблица `publishers`

Содержит данные об издательствах:


* `publisher_id` — идентификатор издательства;
* `publisher` — название издательства.

Таблица `ratings`

Содержит данные о пользовательских оценках книг:

* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.

Таблица `reviews`

Содержит данные о пользовательских обзорах на книги:


* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, написавшего обзор;
* `text` — текст обзора.

__Подключение__

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

## Анализ таблиц

Таблица `books`

In [5]:
query = '''

SELECT *
FROM books
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
query = '''

SELECT COUNT(*)
FROM books

'''

pd.io.sql.read_sql(query, con = engine)

count
0   1000

Таблица `authors`

In [7]:
query = '''

SELECT *
FROM authors
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [8]:
query = '''

SELECT COUNT(*)
FROM authors

'''

pd.io.sql.read_sql(query, con = engine)

count
0    636

Таблица `publishers`

In [9]:
query = '''

SELECT *
FROM publishers
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [10]:
query = '''

SELECT COUNT(*)
FROM publishers

'''

pd.io.sql.read_sql(query, con = engine)

count
0    340

Таблица `ratings`

In [11]:
query = '''

SELECT *
FROM ratings
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [12]:
query = '''

SELECT COUNT(*)
FROM ratings

'''

pd.io.sql.read_sql(query, con = engine)

count
0   6456

Таблица `reviews`

In [13]:
query = '''

SELECT *
FROM reviews
LIMIT 5

'''

pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [14]:
query = '''

SELECT COUNT(*)
FROM reviews

'''

pd.io.sql.read_sql(query, con = engine)

count
0   2793

Были рассмотрены таблицы и количесвто строк в них

## Задания

### Задача №1

Расчет количества книг, вышедших с 1 января 2000 года

In [15]:
query = '''

SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'

'''

pd.io.sql.read_sql(query, con = engine)

count
0    819

Из общей базы в 1000 книг, 819 вышло позже 1 января 2000 года.

### Задача №2

Расчет количества обзоров и средней оценки каждой книги

In [16]:
query = '''

SELECT b.book_id, 
       b.title, 
       COUNT(DISTINCT rv.review_id) AS cnt_reviews, 
       ROUND(AVG(r.rating),3) AS avg_rating
FROM books AS b

LEFT JOIN reviews AS rv
ON b.book_id = rv.book_id

LEFT JOIN  ratings AS r
ON b.book_id = r.book_id

GROUP BY b.book_id, b.title
ORDER BY cnt_reviews DESC

'''

pd.io.sql.read_sql(query, con = engine)

book_id                                              title  cnt_reviews  \
0        948                            Twilight (Twilight  #1)            7   
1        963                                Water for Elephants            6   
2        734                                   The Glass Castle            6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...            6   
4        695  The Curious Incident of the Dog in the Night-Time            6   
..       ...                                                ...          ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel            0   
996      808                            The Natural Way to Draw            0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites            0   
998      221                          Essential Tales and Poems            0   
999      191  Disney's Beauty and the Beast (A Little Golden...            0   

     avg_rating  
0         3.663  
1         3.977  
2         4.207  
3         4.415  
4         4.081  
..          ...  
995       3.667  
996       3.000  
997       5.000  
998       4.000  
999       4.000  

[1000 rows x 4 columns]

Получилась таблица с id  книги, ее названием общим количеством ревью и средней оценкой книги. Количество ревью варируется от 0 до 7 штук.

### Задача №3

Определить издательство, которое выпустило большее число книг толще 50 страниц

In [17]:
query = '''

WITH 
tab AS (
SELECT publisher_id,
       book_id
FROM books
WHERE num_pages > 50
)

SELECT tab.publisher_id,
       p.publisher
FROM tab
INNER JOIN publishers AS p ON tab.publisher_id = p.publisher_id
GROUP BY tab.publisher_id, p.publisher
ORDER BY COUNT(tab.book_id) DESC
LIMIT 1

'''

pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher
0           212  Penguin Books

Издательство Penguin Books выпустило больше всего книг с объемом страниц больше 50-ти.

### Задача №4

Найдем автора с самой высокой средней оценкой книг

In [18]:
query = '''

WITH 

tab AS (
SELECT book_id
FROM ratings
GROUP BY book_id
HAVING COUNT(rating_id) > 50)

SELECT a.author_id, 
       a.author, 
       AVG(rt.rating) AS avg_rating
FROM books AS b

INNER JOIN tab ON b.book_id = tab.book_id

INNER JOIN authors AS a ON b.author_id = a.author_id

INNER JOIN ratings AS rt ON tab.book_id = rt.book_id

GROUP BY a.author_id, a.author
ORDER BY AVG(rt.rating) DESC

LIMIT 1

'''

pd.io.sql.read_sql(query, con = engine)

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.287097

Соавторы J.K. Rowling/Mary GrandPré имеют самую высокую среднюю оценку за свои книги.

### Задача №5

Сосчитаем среднее количество обзоров от пользователей, который поставили больше 48 оценок

In [19]:
query = '''

SELECT COUNT(DISTINCT rv.text)/ COUNT(DISTINCT rv.username) AS avg_reviews_count
FROM reviews AS rv
INNER JOIN ratings AS rt ON rv.username = rt.username
WHERE rt.username IN (SELECT username
                      FROM ratings
                      GROUP BY username
                      HAVING COUNT(rating_id) > 48
                      )

'''

pd.io.sql.read_sql(query, con = engine)

avg_reviews_count
0                 24

В среднем у пользователей, поставивших больше 48 оценок, количество обзоров равно 24 штукам.

## Вывод

Мы получили выводы:

* Из общей базы в 1000 книг, 819 вышло позже 1 января 2000 года.
* Количество ревью варируется от 0 до 7 штук.
* Издательство Penguin Books выпустило больше всего книг с объемом страниц больше 50-ти.
* Соавторы J.K. Rowling/Mary GrandPré имеют самую высокую среднюю оценку за свои книги.
* В среднем у пользователей, поставивших больше 48 оценок, количество обзоров равно 24 штукам.

По этим данным можно сдеать общий вывод, что мы получили довольно новую базу относитеьно современных книг. Самый активный ревьюер оставил 7 обзоров. Самый крупный издатель среди наших данных выпустил 50 книг, а самые популярные авторы - это тандем Джоан Роулинг и Мэри Грандпре. И можно также заключить, что пользователи, которые ставят большое количество оценок, также любят писать обзоры.

## Задача №6

Выведите таблицу, которая будет содержать по году публикации:
* количество издательств,
* выпущенных книг,
* сколько всего тысяч страниц было в изданных книгах

(отобразить только те года, в которых издано более 30 книг)

In [20]:
query = '''

SELECT EXTRACT(YEAR FROM b.publication_date) AS year,
       COUNT(DISTINCT publisher_id) AS publisher_number,
       COUNT(book_id) AS books_number,
       SUM(b.num_pages) / 1000 AS thousand_pages
FROM books AS b
GROUP BY year
HAVING COUNT(book_id) > 30

'''

pd.io.sql.read_sql(query, con = engine)

year  publisher_number  books_number  thousand_pages
0  1999.0                26            41              15
1  2000.0                35            38              13
2  2001.0                41            60              21
3  2002.0                62            94              38
4  2003.0                65           105              41
5  2004.0                88           124              46
6  2005.0                89           139              55
7  2006.0               109           184              68
8  2007.0                38            50              18

Среди всех лет, в которые выпускались книги, только с 1999 по 2007 издавалось больше 30 книг. При этом больше всего страниц написали в 2005 и 2006 годах.

## Задача №7

Выведите в одной таблице два числа — среднюю оценку тех книг, на которые написало отзывов более 3 человек и отдельно среднюю оценку остальных книг, сделайте выводы какой рейтинг больше

In [21]:
query = '''

WITH tab_1 AS (
SELECT b.book_id,
       COUNT(DISTINCT rv.username),
       AVG(rt.rating)
FROM books AS b
JOIN ratings AS rt ON b.book_id = rt.book_id
JOIN reviews AS rv ON rt.book_id = rv.book_id
GROUP BY b.book_id
HAVING COUNT(DISTINCT rv.username) > 3),


tab_2 AS (
SELECT b.book_id,
       COUNT(DISTINCT rv.username),
       AVG(rt.rating)
FROM books AS b
JOIN ratings AS rt ON b.book_id = rt.book_id
JOIN reviews AS rv ON rt.book_id = rv.book_id
GROUP BY b.book_id
HAVING COUNT(DISTINCT rv.username) <= 3)
                     
SELECT (SELECT AVG(avg) AS avg_rating_over_3_review FROM tab_1),
       (SELECT AVG(avg) AS avg_rating_others FROM tab_2)
LIMIT 1


'''

pd.io.sql.read_sql(query, con = engine)

avg_rating_over_3_review  avg_rating_others
0                  3.925045           3.890496

Из полученных данных можно сделать вывод, что книги с более чем тремя ревью, имею более высокий рейтинг, чем все остальные.